src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -168,15 +168,19 @@ std::error_code partition_allocator::check_cluster_limits(
 
     // Refuse to create a partition count that would violate the per-core
     // limit.
-    const uint64_t core_limit = (effective_cpu_count * _partitions_per_shard());
+    const uint64_t core_limit = (effective_cpu_count * _partitions_per_shard())
+                                - (broker_count * _partitions_reserve_shard0());
```

Comment:
why do we multiply it by `broker_count`?

Reply:
`core_limit` is the max number of partitions (including replicas) that can be created across the cluster, and each broker has a shard0.

> [topic_partitions_reserve_shard0](https://docs.redpanda.com/current/reference/properties/cluster-properties/#topic_partitions_reserve_shard0): Reserved partition slots on shard (CPU core) 0 on each node. If this is greater than or equal to [topic_partitions_per_core](https://docs.redpanda.com/current/reference/properties/cluster-properties/#topic_partitions_per_core), no data partitions will be scheduled on shard 0.

Reply:
Thanks, makes sense.

TBH I would find it more readable if we restructured `core_limit` calculation as `broker_count * (min_core_count * _partitions_per_shard() - _partitions_reserve_shard0())`. Not pushing, you do you.

Oh and there seems to be a typo in the piece of the docs you cited: it should be `topic_partitions_per_shard`, not `topic_partitions_per_core`, right?

Reply:
Looks like the docs part is already fixed: https://github.com/redpanda-data/redpanda/pull/22893